### PR TITLE
DANet

### PR DESCRIPTION
**Pull request by xdeng7**
Add DANet implemented with attention

### CODE REVIEW

**Comment by xdeng7** [comment]
Fix segbase and loss

**Comment by mli** [comment]
Job PR-1280-17 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1280/17/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/loss.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 __all__ = ['FocalLoss', 'SSDMultiBoxLoss', 'YOLOV3Loss',
            'MixSoftmaxCrossEntropyLoss', 'ICNetLoss', 'MixSoftmaxCrossEntropyOHEMLoss',
+           'SegmentationMultiLosses',
```
Comment (bryanyzhu) [inline gluoncv/loss.py:12]: You can write multiple items in one line, try to make the code look nicer.

File: gluoncv/loss.py [hunk 2]
```diff
@@ -287,6 +288,24 @@ def hybrid_forward(self, F, pred, label):
...
         return F.mean(loss, axis=self._batch_axis, exclude=True)
 
+
+class SegmentationMultiLosses(SoftmaxCrossEntropyLoss):
+    """2D Cross Entropy Loss with Multi-Loss"""
+    def __init__(self, size_average=True, ignore_label=-1, **kwargs):
+        super(SegmentationMultiLosses, self).__init__(size_average, ignore_label, **kwargs)
+
+    def hybrid_forward(self, F, *inputs, **kwargs):
+        # *preds, label = tuple(inputs)
```
Comment (bryanyzhu) [inline gluoncv/loss.py:298]: Remove commented lines after debugging.

File: gluoncv/model_zoo/attention.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+from mxnet.gluon import nn
+from mxnet.gluon.nn import HybridBlock
+# pylint: disable-all
+
+__all__ = ['PAM_Module', 'CAM_Module']
+
+
+class PAM_Module(HybridBlock):
+    """ Position attention module"""
+    #Ref from SAGAN
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/attention.py:11]: Add more descriptions on what is a PAM module and a CAM module. If this is proposed in a paper, please put the paper title and link here.

File: gluoncv/model_zoo/attention.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        out = F.batch_dot(proj_value, attention, transpose_b=True)
+        out = F.reshape_like(out, x, lhs_begin=2, lhs_end=None, rhs_begin=2, rhs_end=None)
+
+        out = gamma * out + x
+
+        return out
+
+
+class CAM_Module(HybridBlock):
+    """ Channel attention module"""
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/attention.py:44]: Same thing, add more descriptions and reference.

File: gluoncv/model_zoo/danet.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        'resnet101').
+    norm_layer : object
+        Normalization layer used in backbone network (default: :class:`mxnet.gluon.nn.BatchNorm`;
+    Reference:
+        Jun Fu, Jing Liu, Haijie Tian, Yong Li, Yongjun Bao, Zhiwei Fang, Hanqing Lu. "Dual Attention 
+        Network for Scene Segmentation." *CVPR*, 2019
+    """
+
+    def __init__(self, nclass, backbone='resnet50', aux=False, ctx=cpu(), pretrained_base=True,
+                 height=None, width=None, base_size=576, crop_size=512, dilated=True, **kwargs):
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/danet.py:29]: Why base_size 576 and crop_size 512? It is kind of strange, because for ADE, the default is 480 and for cityscapes, the default is 768.

File: gluoncv/model_zoo/danet.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        'coco': 'coco',
+        'citys': 'citys',
+    }
+    from ..data import datasets
+    # infer number of classes
+    model = DANet(nclass=datasets[dataset].NUM_CLASS, backbone=backbone, ctx=ctx, **kwargs)
+    model.classes = datasets[dataset].classes
+    if pretrained:
+        from .model_store import get_model_file
+        model.load_parameters(get_model_file('danet_%s_%s'%(backbone, acronyms[dataset]),
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/danet.py:161]: add space before  and after %

File: gluoncv/model_zoo/model_zoo.py [hunk 7]
```diff
@@ -184,6 +185,8 @@
     'deeplab_v3b_plus_wideresnet_citys': get_deeplab_v3b_plus_wideresnet_citys,
     'icnet_resnet50_citys': get_icnet_resnet50_citys,
     'icnet_resnet50_mhpv1': get_icnet_resnet50_mhpv1,
+    'danet_resnet50_citys':get_danet_resnet50_citys,
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/model_zoo.py:188]: Add one space after :

File: scripts/segmentation/train.py [hunk 8]
```diff
@@ -109,7 +109,7 @@ def parse_args():
         args.ctx = [mx.cpu(0)]
     else:
         print('Number of GPUs:', args.ngpus)
-        assert args.ngpus > 0, 'No GPUs found, please enable --no-cuda for CPU mode.'
+        # assert args.ngpus > 0, 'No GPUs found, please enable --no-cuda for CPU mode.'
```
Comment (bryanyzhu) [inline scripts/segmentation/train.py:112]: Please add this line back, we need this assertion.

File: scripts/segmentation/train.py [hunk 9]
```diff
@@ -192,6 +192,8 @@ def __init__(self, args, logger):
         # create criterion
         if 'icnet' in args.model:
             criterion = ICNetLoss(crop_size=args.crop_size)
+        elif 'danet' in args.model or 'danet' in args.model_zoo:
+            criterion=SegmentationMultiLosses()
```
Comment (bryanyzhu) [inline scripts/segmentation/train.py:196]: Add space before and after "="

Comment (xdeng7) [reply scripts/segmentation/train.py]: Thanks for your comments. I've made the changes based on them.

**Comment by mli** [comment]
Job PR-1280-18 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1280/18/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-1280-20 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1280/20/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-1280-19 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1280/19/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-1280-22 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1280/22/index.html 
Code coverage of this PR:  vs. Master:

**Comment by bryanyzhu** [comment]
PR looks good to me, looking forward to the trained models/logs. I will then merge it.

**Comment by xdeng7** [comment]
> PR looks good to me, looking forward to the trained models/logs. I will then merge it.

Thank you. Sounds good.

**Comment by mli** [comment]
Job PR-1280-29 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1280/29/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-1280-30 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1280/30/index.html 
Code coverage of this PR:  vs. Master:
